config/default/configmap/inferenceservice.yaml

```diff
@@ -70,6 +70,10 @@ data:
             ]
         }
     }
+  deployment: |-
+    {
+        "queueSideCarResourcePercentage": "0.1"
```

Comment:
I'm not sold on us needing to expose this via our configmap. Isn't this something that can be configured in the Knative configmap instead? Happy to hear arguments, but my initial impression is that this isn't an ML specific problem.

Reply:
+1 to this feels like a Knative thing... Can possibly be the job of a separate admission controller on a KNative service - that would enable/help anyone who is using Knative to apply a resource quota (independent of KFServing)?

Reply:
It seems to me that we have two ways to fix the queue_proxy issue if we do not want to expose this
option1: Set a default value in our code but do not expose it in the configmap lol
option2: Do not set CPU/Memory Limit/Request for queue_proxy from kfserving. Instead, user should run a mutation webhook service in their cluster that checks all the kfserving pods and inject the CPU/Memory configu to the queue_proxy container. 

what do you guys think?

Reply:
I dont think you need to mutate pods, adding the annotation to the KN service (or even KF service) should be enough? Cleanest is to have a mutator for KNService (as that benefits the entire KN community - maybe this config should live there?)

That said, another option to consider is to model this more generically as a config of a set of annotations that get added to every incoming KFService? This would then be an operator hook (operator being the person who sets up the cluster) that allows the operator to add whatever arbitrary annotations they need for their KF deployment... (and the knowledge of this specific annotation is not built into our controller).

wdyt @ellis-bigelow

Reply:
Not sure I'm excited about either. Perhaps a configmap values in knative. What about https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/memory-default-namespace/#create-a-limitrange-and-a-pod. Does that work?

Reply:
@ellis-bigelow this is the first option I proposed to our team and got rejected :( Multiple team members didn't like to apply a default CPU/Memory Limit/Request to namespaces because we want to know which containers are missing this type configuration. We are also concerned that the default configuration may be too small for some containers and they will just OOMed.

Reply:
I no longer expose the queueSideCarResourcePercentage configuration from the configmap but still set it to 0.1 in the code. Would this be better :)

Reply:
I agree -- this may not be something you want to leak this to your entire namespace. When using Istio, how are you guaranteeing the minimum sidecar resources?

I'm willing to accept that we set a default resource request for all containers we configure. I want to understand why we can't configure this at a knative level via a configmap or otherwise. I see they have the annotation which enables this, but it might be nice to have it more global.

Another option is to simply put this annotation that you expect on the knative resources onto your KFServing InferenceServices. We pass all of our annotations down to the underlying resources, so you can get this feature that way.

I could see us allowing users to set the storage initializer defaults in our configmap, the same way that the knative queue proxy defaults should be set in the knative configmap.

I'm trying to optimize for:
- push all possible dependency changes down into dependencies
- enable customers any configurability they need
- allow users to specify global defaults with configmaps.

Reply:
Are you suggesting make Knative to have a global config for the queue-proxy resources like the storage initializer defaults we did here? We definitely do not want user to set this annotation on InferenceService themselves, users do not understand what queue-proxy and they have no idea how much resources are needed.

Reply:
@ellis-bigelow I am glad that you agree that kfserving should set a default resource request for all containers. And I am setting a default resource request for queue_proxy in the kfserving code here: https://github.com/yuzliu/kfserving/blob/c5bacf4029fb296862926ef6d12541732a357b6d/pkg/controller/inferenceservice/resources/knative/service.go#L46 and I am no longer exposing this from kfserving's configmap since I do agree that kfserving's users shouldn't worry about setting this value. If they really want to override this annotation they can do it in their own annotation service. 

About your quetion why shouldn't we set this default value in knative? I think it will be nice if we could do that and I raised an issue for knative serving folks: https://github.com/knative/serving/issues/5829 :) But tbh it is way easier if we can set a default value here and start to run kfserving in Prod :)

Reply:
I'm convinced! Thanks for your patience while we figured out the best path forward here, Yuzliu!